### DISCUSSION TITLE
Using modbus-proxy

### DISCUSSION THREAD

**Discussion by javierin** [basking-in-the-sun2000/solar-logger#19] [Ideas]
As you may already have suffered, SUN2000 easily gets stuck when more than a single client connects to its TCP Modbus interface. Damn. Sometimes it hangs with just a single client. 

I have found a [modbus-proxy software](https://github.com/tiagocoutinho/modbus-proxy) which should help with the task of accessing the inverter from several different devices (solar logger, zabbix, homeassistant, etc)
However, I cannot manage to get it working and I have no clue of where to start investigating.

Has anybody else tried this modbus proxy with the SUN2000 with success?

**Comment by Roving-Ronin** [comment]
Wiring inverter(s) and DTSU-666H meter to Huawei Data Logger would solve this issue, with all data pulled into the logger via RS485 and then made available via Modbus/TCP. 

If only it was possible to work out the damn register ID's for access this data from the logger though.

**Reply by basking-in-the-sun2000** [reply]
Take a look at this [thread](https://forum.huawei.com/enterprise/en/modbus-definitions-for-dtsu666-h-for-my-own-datalogger-on-modbus-rtu/thread/672405-100027). You either reply to the meter registers, or work as a repeater transferring data between both modbus ports. Modbus isn't a very smart protocol, and there can only be one master on each bus

However, don't think the inverter would pull the data through tcp. Huawei's dataloggers x000 serve as repeaters, reading the meter and sending the data to the inverter through rtu. 

Initially, I thought of going this route and having the logger work as a repeater. However, I like blocking tcp and think of it as a soft security feature. I know disconnecting the logger is a pain, but I don't do it too often. Also not a fan of adding a power device to my network. 

Just pushed some changes that should allow using either tcp or rtu modbus. Since I don't have a rtu connection, I didn't work out the code, but the ground is set to use pymodbus rtu connections.
